File: ConsoleApplication9/ConsoleApplication9/IterCLAY.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ConsoleApplication9
{
    public static class IterSlAY
    {
        public delegate double _baseFunction(double item1, double item2);

        public static double MethodDeterminatFunc(_baseFunction[,] baseFunction, double[,] baseValue)
        {
            double[,] newMatrix = new double[baseFunction.GetLength(0), baseFunction.GetLength(1)];
            for (UInt16 i = 0; i < newMatrix.GetLength(0); ++i)
                for (UInt16 j = 0; j < newMatrix.GetLength(1); ++j)
                {
                    newMatrix[i, j] = baseFunction[i, j](baseValue[0, 0], baseValue[1, 0]);
                }
            return Determinat.Determinat_Method(newMatrix);
        }

        public static double[,] MethodDecision(double[,] baseX, double accyra)
        {
            _baseFunction[,] iakobi = new _baseFunction[2, 2];

            iakobi[0, 0] = (double x, double y) => y;
            iakobi[0, 1] = (double x, double y) => x;
            iakobi[1, 0] = (double x, double y) => 2 * x;
            iakobi[1, 1] = (double x, double y) => 2 * y;

            _baseFunction[,] delta1 = new _baseFunction[2, 2];

            delta1[0, 0] = (double x, double y) => x * y -1;
            delta1[0, 1] = (double x, double y) => x;
            delta1[1, 0] = (double x, double y) => y*y+x*x-4;
            delta1[1, 1] = (double x, double y) => 2*y;

            _baseFunction[,] delta2 = new _baseFunction[2, 2];

            delta2[0, 0] = (double x, double y) => y;
            delta2[0, 1] = (double x, double y) => x*y-1;
            delta2[1, 0] = (double x, double y) => 2 * x;
            delta2[1, 1] = (double x, double y) => y * y + x * x - 4;

            return RecursionDesicion(iakobi, delta1, delta2, baseX, accyra);
        }

        private static double[,] RecursionDesicion(_baseFunction[,] iakobi,
            _baseFunction[,] delta1, _baseFunction[,] delta2, double[,] baseX, double accyra)
        {
            double iakobiD = MethodDeterminatFunc(iakobi, baseX);
            double delta1D = MethodDeterminatFunc(delta1, baseX);
            double delta2D = MethodDeterminatFunc(delta2, baseX);

            double[,] newX = new double[2, 1];
            newX[0, 0] = baseX[0, 0] - delta1D / iakobiD;
            newX[1, 0] = baseX[1, 0] - delta2D / iakobiD;

            if (Math.Abs(newX[0, 0] - baseX[0, 0]) > accyra || Math.Abs(newX[1, 0] - baseX[1, 0]) > accyra)
                newX = RecursionDesicion(iakobi, delta1, delta2, newX, accyra);
            return newX;
        }
    }
}

File: Task/3/IterCLAY.py
import math, Matrix
import numpy as np

#Вычисления массива функций
def DeterminatFunc(baseFunc,baseValue):
    newValue = np.zeros((baseFunc.shape[0],baseFunc.shape[1]))

    for i in range(0,newValue.shape[0]):
        for j in range(0,newValue.shape[1]):
            newValue[i][j] = baseFunc[i][j](baseValue[0],baseValue[1])

    return Matrix.Determinat(newValue)

#Метод задания матриц для задания 3.2
def Decision1(baseX, accur):
    ikobi = np.array([[lambda x,y: 2*x,     lambda x,y:  2*y],
                              [lambda x,y: 3*x**2, lambda x,y:  -1   ]])

    delta1 = np.array([[lambda x,y: x**2 + y**2 - 1, lambda x,y:  2*y],
                                [lambda x,y: x**3 - y,          lambda x,y:  -1   ]])

    delta2 = np.array([[lambda x,y: 2*x,     lambda x,y:  x**2 + y**2 - 1],
                                 [lambda x,y: 3*x**2, lambda x,y:  x**3 - y  ]])

    return RecursionDesicion(ikobi,delta1,delta2,baseX,accur)

#Метод задания матриц для задания 3.3
def Decision2(baseX, accur):
    ikobi = np.array([[lambda x,y: y/(math.cos(x*y+0.4)**2) -2*x,     lambda x,y:  x/(math.cos(x*y+0.4)**2)],
                              [lambda x,y: 1.2*x, lambda x,y:  4*y  ]])

    delta1 = np.array([[lambda x,y: math.tan(x*y+0.4)-x*x,     lambda x,y:  x/(math.cos(x*y+0.4)**2)],
                              [lambda x,y: 0.6*x*x+2*y*y-1, lambda x,y:  4*y  ]])

    delta2 = np.array([[lambda x,y: y/(math.cos(x*y+0.4)**2) -2*x,     lambda x,y: math.tan(x*y+0.4)-x*x],
                              [lambda x,y: 1.2*x, lambda x,y:  0.6*x*x+2*y*y-1 ]])

    return RecursionDesicion(ikobi,delta1,delta2,baseX,accur)

#Рекурсионный метод итерационнго вычисления
def RecursionDesicion(ikobi,delta1,delta2,baseX,accur):
    
    ikobiD = DeterminatFunc(ikobi,baseX)
    delta1D = DeterminatFunc(delta1,baseX)
    delta2D = DeterminatFunc(delta2,baseX)

    newX = (baseX[0]-delta1D/ikobiD,baseX[1]-delta2D/ikobiD)

    if abs(newX[0] - baseX[0]) > accur or abs(newX[1] - baseX[1]) > accur:
        newX = RecursionDesicion(ikobi,delta1,delta2,newX,accur)
    return newX

x = 1
y = 1

accuracy = 0.0001

print("Решения",Decision1((1,1),accuracy))

x = 1
y = 1

accuracy = 0.001

print("Решения",Decision2((1,1),accuracy))
